amaranth/sim/core.py

```diff
@@ -264,4 +265,8 @@ def write_vcd(self, vcd_file, gtkw_file=None, *, traces=()):
                     file.close()
             raise ValueError("Cannot start writing waveforms after advancing simulation time")
 
+        for trace in traces:
+            if isinstance(trace, Signal) and trace.name == "":
+                raise TypeError("Cannot trace private Signal")
```

Comment:
I think this isn't sufficient due to how pysim VCD writer accepts a bunch of other stuff too, like value-castables, in `traces`.

Reply:
I copied the logic from pysim, it seems more reasonable and I was able to test it.